### PR TITLE
268 add endpoint for party unlocking

### PR DESCRIPTION
**Pull request by shahi645**
# Motivation and Context
When an account was unlocked, only django can unlock user, so have added unlocking ability that can be called from party service 

# What has changed
Unlock-ability, previously only capable from the DB

# How to test?
Run in conjunction with other branches:
Frontstage: `268-update-party-service-with-acount-lock`
ROps: `268-ability-to-unlock-a-respondents-account`
Party: `268-party-lock-account-in-djano-oauth2`

Try unlocking from ROps, should work and be able to log in. Also all other aspects aroun OAuth2 should remain same

### CODE REVIEW

File: oauth2server/apps/accounts/decorators.py [hunk 1]
```diff
@@ -148,25 +147,25 @@ def _extract_username(request, **kwargs):
...
             except KeyError:
                 stdlogger.warning("Username is missing from the PUT method")
                 raise UsernameRequiredException()
             except ValidationError:
                 stdlogger.warning("Email failed validation check for validating a user")
                 raise InvalidUserCredentialsException
             except OAuthUser.DoesNotExist:
-                stdlogger.error("Error while updating the User ID. The email: {} does not exist on the OAuth2 server".format(username))
+                stdlogger.error(
+                    "Error while updating the User ID. The email: {} does not exist on the OAuth2 server".format(
```
Comment (atrozzo) [inline oauth2server/apps/accounts/decorators.py]: someone was pointing out that email should not be logged ? I find teh information useful but not sure is complaiant with ONS internal rules ? @gavinleeedwards

File: oauth2server/apps/accounts/views.py [hunk 2]
```diff
@@ -81,10 +80,28 @@ def put(self, request):
...
+                stdlogger.info('Admin is unlocking account')
+                request.user.unlock_account()
+                request.user.reset_failed_logins()
+                # Verify user if not verified
+                if not request.user.account_is_verified:
+                    request.user.verify_account()
+            request.user.save()
+
+        except (IntegrityError, InternalError, DataError, DatabaseError):
+            # The chances of this happening are slim to none! And this line of code should never happen. So we really
```
Comment (atrozzo) [inline oauth2server/apps/accounts/views.py]: I would actually log the error.

File: oauth2server/apps/accounts/views.py [hunk 3]
```diff
@@ -81,10 +80,28 @@ def put(self, request):
             # need to tell the other system we are not capable of creating the resource.
             raise DatabaseFailureException
 
+        # if account_locked attribute set to False, we want to unlock the account
+        try:
+            # Check to see if the POST contains the account_locked flag.
+            # If so we unlock account and reset failed log-ins.
+            if request.POST.get('account_locked') == 'False':
+                stdlogger.info('Admin is unlocking account')
```
Comment (atrozzo) [inline oauth2server/apps/accounts/views.py]: this looks like more a debug level info.

File: oauth2server/apps/accounts/views.py [hunk 4]
```diff
@@ -81,10 +80,28 @@ def put(self, request):
...
+                request.user.unlock_account()
```
Comment (atrozzo) [inline oauth2server/apps/accounts/views.py]: any chance any of these 2 can fail ?

File: oauth2server/apps/accounts/tests/test_user_admin.py [hunk 5]
```diff
@@ -1,34 +1,36 @@
...
     fixtures = ['test_credentials',]
 
     def setUp(self):
         self.api_client = APIClient()
 
-    # Make sure that client_id with admin privilege can't update the account if it's in a locked state.
+    # Make sure that client_id with admin privilege can update the account if it's in a locked state.
     # Use john3@doe.com which has a locked account
-    def test_locked_account(self):
+    def test_unlock_locked_account(self):
```
Comment (atrozzo) [inline oauth2server/apps/accounts/tests/test_user_admin.py:18]: This cover just the positive test. You stil miss : 
1.          except (IntegrityError, InternalError, DataError, DatabaseError): 
2.         if not request.user.account_is_verified:

**Review by atrozzo — CHANGES_REQUESTED** [review]

File: oauth2server/apps/accounts/decorators.py (hunk 1 above)
Comment (shahi645) [reply oauth2server/apps/accounts/decorators.py]: It was already there, and i don't think this goes to anything public facing?

File: oauth2server/apps/accounts/views.py (hunk 2 above)
Comment (shahi645) [reply oauth2server/apps/accounts/views.py]: I think this is original code, can happily change but out of scope so need go ahead

File: oauth2server/apps/accounts/views.py (hunk 4 above)
Comment (shahi645) [reply oauth2server/apps/accounts/views.py]: Nope they are in the model

File: oauth2server/apps/accounts/decorators.py (hunk 1 above)
Comment (tej99) [inline oauth2server/apps/accounts/decorators.py]: Email address is sensitive data that shouldn't be logged.

File: oauth2server/apps/accounts/decorators.py [hunk 6]
```diff
@@ -175,7 +174,9 @@ def _extract_username(request, **kwargs):
                 validate_email(new_username)
                 check_user = OAuthUser.objects.get(email=new_username)
                 if check_user:
-                    stdlogger.warning("The user: {} tried to change their email to the already active user: {}".format(username, new_username))
+                    stdlogger.warning(
+                        "The user: {} tried to change their email to the already active user: {}".format(username,
```
Comment (tej99) [inline oauth2server/apps/accounts/decorators.py]: Email address is sensitive data that shouldn't be logged.

File: oauth2server/apps/accounts/decorators.py [hunk 7]
```diff
@@ -212,7 +213,8 @@ def _extract_username(request, **kwargs):
                 raise InvalidUserCredentialsException
             except OAuthUser.DoesNotExist:
                 stdlogger.error(
-                    "Error while getting the User ID. The email: {} does not exist on the OAuth2 server".format(username))
+                    "Error while getting the User ID. The email: {} does not exist on the OAuth2 server".format(
```
Comment (tej99) [inline oauth2server/apps/accounts/decorators.py]: Email address again

File: oauth2server/apps/accounts/decorators.py [hunk 8]
```diff
@@ -229,14 +231,13 @@ def _extract_username(request, **kwargs):
                 raise InvalidUserCredentialsException
             except OAuthUser.DoesNotExist:
                 stdlogger.error(
-                    "Error while trying to DELETE the User. The email: {} does not exist on the OAuth2 server".format(username))
+                    "Error while trying to DELETE the User. The email: {} does not exist on the OAuth2 server".format(
```
Comment (tej99) [inline oauth2server/apps/accounts/decorators.py]: Email in log

File: oauth2server/apps/accounts/views.py [hunk 9]
```diff
@@ -81,10 +80,28 @@ def put(self, request):
             # need to tell the other system we are not capable of creating the resource.
             raise DatabaseFailureException
 
+        # if account_locked attribute set to False, we want to unlock the account
+        try:
+            # Check to see if the POST contains the account_locked flag.
+            # If so we unlock account and reset failed log-ins.
+            if request.POST.get('account_locked') == 'False':
```
Comment (tej99) [inline oauth2server/apps/accounts/views.py]: Can we not merge both try excepts into one? Probably only need the try except around the `request.user.save()`

File: oauth2server/apps/accounts/views.py [hunk 10]
```diff
@@ -81,10 +80,28 @@ def put(self, request):
...
+                stdlogger.debug('Admin is unlocking account')
+                request.user.unlock_account()
+                request.user.reset_failed_logins()
+                # Verify user if not verified
+                if not request.user.account_is_verified:
```
Comment (tej99) [inline oauth2server/apps/accounts/views.py:81]: Not sure this is right. What if an unverified account is locked? if you unlock it you wouldn't want to automatically verify the account as well.

**Review by tej99 — CHANGES_REQUESTED** [review]
Tested with party and it unlocks the user successfully 👍 But got a few questions and change requests

File: oauth2server/apps/accounts/decorators.py [hunk 11]
```diff
@@ -53,15 +53,14 @@ def _extract_client(request):
...
-
         # First, let's check Authorization header if present
         # TODO - Look at removing this and encrypting password. It's an OAuth2 server, so it should be over SSL anyway
         # but revisit this if time permits.
         if 'HTTP_AUTHORIZATION' in request.META:
             stdlogger.debug("We have a HTTP_AUTHORIZATION request ***")
             auth_header = request.META['HTTP_AUTHORIZATION']
             auth_method, auth = re.split(':|;|,| ', auth_header)
-            #auth_method, auth = request.META['HTTP_AUTHORIZATION'].split(':')
+            # auth_method, auth = request.META['HTTP_AUTHORIZATION'].split(':')
```
Comment (MIslam92) [inline oauth2server/apps/accounts/decorators.py]: Remove this comment?

File: oauth2server/apps/accounts/decorators.py (hunk 1 above)
Comment (shahi645) [reply oauth2server/apps/accounts/decorators.py]: @tej99 This was already there so not sure whether to remove but i guess i can remove?

File: oauth2server/apps/accounts/views.py (hunk 10 above)
Comment (shahi645) [reply oauth2server/apps/accounts/views.py:81]: We have had approval from the business that this is the behaviour we want, as we are happy at this point they have either:

- access their email to reset (i.e. unlock)
- Called up internal and verified over the phone

File: oauth2server/apps/accounts/decorators.py (hunk 1 above)
Comment (tej99) [reply oauth2server/apps/accounts/decorators.py]: Yes best to remove it

**Review by tej99 — APPROVED** [review]

File: oauth2server/apps/accounts/decorators.py [hunk 12]
```diff
@@ -323,13 +315,10 @@ def decorator(request, *args, **kwargs):
         _extract_client(request=request)
         _extract_username(request=request, *args, **kwargs)
         _extract_active(request=request)
-        #_extract_scope(request=request)
-        #_extract_account_locked(request=request)
-        #_extract_failed_login(request=request)
+        # _extract_scope(request=request)
```
Comment (MIslam92) [inline oauth2server/apps/accounts/decorators.py]: We're not using these? If not remove?

**Review by MIslam92 — CHANGES_REQUESTED** [review]

**Review by MIslam92 — APPROVED** [review]

**Pull request merged**